File: 394-Decode String.py
# stack
class Solution(object):
    def decodeString(self, s):
        stack, num, string = [], 0, ''
        for char in s:
            if char == '[':  # 开始，进栈
                stack.append(string)
                stack.append(num)
                string, num = '', 0
            elif char == ']':  # 结束，出栈
                num = stack.pop()
                pre_string = stack.pop()
                string = pre_string + num * string
                num = 0
            elif char.isdigit():  # 是数字
                num = num * 10 + int(char)
            else:  # 字符串
                string += char
            print(string, stack, num)
        return string

File: C++/394-Decode_String.cpp
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Decode String.
// Memory Usage: 8.8 MB, less than 94.12% of C++ online submissions for Decode String.
class Solution {
public:
    string decodeString(string s) {
        stack<string> stack1;
        stack<int> stack2;
        string res;
        int num = 0;
        for (char ch : s) {
            if (ch == '[') {
                stack1.push(res);
                stack2.push(num);
                res = "";
                num = 0;
            } else if (ch == ']') {
                string temp = res;
                for (int i = 0; i < stack2.top() - 1; ++i)
                    res += temp;
                res = stack1.top() + res;
                stack2.pop();
                stack1.pop();
            } else if (isdigit(ch)) {
                num = num * 10 + (ch - '0');
            } else {
                res.push_back(ch);
            }
        }
        return res;
    }
};
